File: android/src/main/java/expo/modules/smartrefreshlayout/FooterPropsHandler.kt
package expo.modules.smartrefreshlayout

import android.content.res.Resources
import android.view.View
import android.view.ViewGroup
import androidx.core.graphics.toColorInt
import com.scwang.smart.refresh.footer.ClassicsFooter
import com.scwang.smart.refresh.layout.SmartRefreshLayout
import com.scwang.smart.refresh.layout.constant.RefreshState as NativeRefreshState

/**
 * Footer 属性处理器
 * 负责处理经典样式 Footer 的所有属性设置
 */
class FooterPropsHandler(
    private val srl: SmartRefreshLayout,
    private val resources: Resources
) {
    
    /**
     * 应用 Footer 属性
     */
    fun applyProps(props: Map<String, Any?>?) {
        props ?: return
        
        val footer = srl.refreshFooter
        if (footer !is ClassicsFooter) {
            return
        }
        
        try {
            // 设置文字内容
            setTextContent(footer, props)
            
            // 强制刷新 Footer 显示
            footer.onStateChanged(srl, NativeRefreshState.None, srl.state)
            
            // 调整 TextView 宽度
            adjustTextViewWidth(footer)
            
            // 设置颜色
            setColors(footer, props)
            
            // 设置字体大小
            setTextSize(footer, props)
            
            // 设置其他属性
            setOtherProps(footer, props)
            
            // 设置箭头大小
            setArrowSize(footer, props)
            
            // 设置间距
            setMargin(footer, props)
        } catch (_: Throwable) {
            // 静默处理异常
        }
    }
    
    /**
     * 设置文字内容
     */
    private fun setTextContent(footer: ClassicsFooter, props: Map<String, Any?>) {
        val textFields = mapOf(
            "REFRESH_FOOTER_PULLING" to "mTextPulling",
            "REFRESH_FOOTER_LOADING" to "mTextLoading",
            "REFRESH_FOOTER_REFRESHING" to "mTextRefreshing",
            "REFRESH_FOOTER_RELEASE" to "mTextRelease",
            "REFRESH_FOOTER_FINISH" to "mTextFinish",
            "REFRESH_FOOTER_FAILED" to "mTextFailed",
            "REFRESH_FOOTER_NOTHING" to "mTextNothing"
        )
        
        textFields.forEach { (propKey, fieldName) ->
            props[propKey]?.let { value ->
                if (value is String) {
                    try {
                        val field = ClassicsFooter::class.java.getDeclaredField(fieldName)
                        field.isAccessible = true
                        field.set(footer, value)
                        
                        // 同时修改静态常量
                        when (propKey) {
                            "REFRESH_FOOTER_PULLING" -> 
                                com.scwang.smart.refresh.footer.ClassicsFooter.REFRESH_FOOTER_PULLING = value
                            "REFRESH_FOOTER_LOADING" -> 
                                com.scwang.smart.refresh.footer.ClassicsFooter.REFRESH_FOOTER_LOADING = value
                            "REFRESH_FOOTER_REFRESHING" -> 
                                com.scwang.smart.refresh.footer.ClassicsFooter.REFRESH_FOOTER_REFRESHING = value
                            "REFRESH_FOOTER_RELEASE" -> 
                                com.scwang.smart.refresh.footer.ClassicsFooter.REFRESH_FOOTER_RELEASE = value
                            "REFRESH_FOOTER_FINISH" -> 
                                com.scwang.smart.refresh.footer.ClassicsFooter.REFRESH_FOOTER_FINISH = value
                            "REFRESH_FOOTER_FAILED" -> 
                                com.scwang.smart.refresh.footer.ClassicsFooter.REFRESH_FOOTER_FAILED = value
                            "REFRESH_FOOTER_NOTHING" -> 
                                com.scwang.smart.refresh.footer.ClassicsFooter.REFRESH_FOOTER_NOTHING = value
                        }
                    } catch (_: Exception) {
                        // 静默处理异常
                    }
                }
            }
        }
    }
    
    /**
     * 调整 TextView 宽度，确保文字完整显示
     */
    private fun adjustTextViewWidth(footer: ClassicsFooter) {
        try {
            val possibleNames = listOf("mTitleText", "mTextTitle", "titleText", "mTitle")
            var titleTextView: android.widget.TextView? = null
            
            // 遍历类层次结构查找 TextView
            var searchClass: Class<*>? = ClassicsFooter::class.java
            outerLoop@ while (searchClass != null && searchClass != Any::class.java) {
                for (fieldName in possibleNames) {
                    try {
                        val field = searchClass.getDeclaredField(fieldName)
                        field.isAccessible = true
                        val view = field.get(footer)
                        if (view is android.widget.TextView) {
                            titleTextView = view
                            break@outerLoop
                        }
                    } catch (_: NoSuchFieldException) {}
                }
                searchClass = searchClass.superclass
            }
            
            if (titleTextView != null) {
                // 调整父容器宽度
                val parent = titleTextView.parent
                if (parent is ViewGroup) {
                    val parentParams = parent.layoutParams
                    if (parentParams != null) {
                        parentParams.width = ViewGroup.LayoutParams.MATCH_PARENT
                        parent.layoutParams = parentParams
                    }
                }
                
                // 设置 TextView 为 WRAP_CONTENT
                val params = titleTextView.layoutParams
                if (params != null) {
                    params.width = ViewGroup.LayoutParams.WRAP_CONTENT
                    titleTextView.layoutParams = params
                }
                
                // 禁用 ellipsize，让文字完整显示
                titleTextView.ellipsize = null
                titleTextView.setSingleLine(true)
                titleTextView.maxLines = 1
                titleTextView.maxWidth = Int.MAX_VALUE
                titleTextView.maxEms = Int.MAX_VALUE
                titleTextView.requestLayout()
            } else {
                // 通过遍历 View 树查找
                titleTextView = findTextViewInTree(footer.view)
                if (titleTextView != null) {
                    val params = titleTextView.layoutParams
                    if (params != null) {
                        params.width = ViewGroup.LayoutParams.WRAP_CONTENT
                        titleTextView.layoutParams = params
                    }
                    titleTextView.setSingleLine(true)
                    titleTextView.maxLines = 1
                    titleTextView.maxWidth = Int.MAX_VALUE
                }
            }
        } catch (_: Exception) {
            // 静默处理异常
        }
    }
    
    /**
     * 在 View 树中查找 TextView
     */
    private fun findTextViewInTree(view: View): android.widget.TextView? {
        if (view is android.widget.TextView && view.text.toString().isNotEmpty()) {
            return view
        }
        if (view is ViewGroup) {
            for (i in 0 until view.childCount) {
                val found = findTextViewInTree(view.getChildAt(i))
                if (found != null) return found
            }
        }
        return null
    }
    
    /**
     * 设置颜色
     */
    private fun setColors(footer: ClassicsFooter, props: Map<String, Any?>) {
        props["footerAccentColor"]?.let { 
            if (it is String) {
                footer.setAccentColor(it.toColorInt())
            }
        }
        props["footerPrimaryColor"]?.let { 
            if (it is String) {
                footer.setPrimaryColor(it.toColorInt())
            }
        }
    }
    
    /**
     * 设置字体大小
     */
    private fun setTextSize(footer: ClassicsFooter, props: Map<String, Any?>) {
        props["footerTitleTextSize"]?.let { 
            if (it is Number) {
                val spValue = it.toFloat()
                footer.setTextSizeTitle(spValue)
            }
        }
    }
    
    /**
     * 设置其他属性
     */
    private fun setOtherProps(footer: ClassicsFooter, props: Map<String, Any?>) {
        props["footerFinishDuration"]?.let { 
            if (it is Number) {
                footer.setFinishDuration(it.toInt())
            }
        }
    }
    
    /**
     * 设置箭头大小
     */
    private fun setArrowSize(footer: ClassicsFooter, props: Map<String, Any?>) {
        val arrowSize = props["footerDrawableArrowSize"] ?: props["footerDrawableSize"]
        arrowSize?.let { 
            if (it is Number) {
                val dpValue = it.toDouble()
                val pxValue = dpToPx(dpValue).toInt()
                
                if (dpValue == 0.0) {
                    // 隐藏箭头
                    hideArrows(footer)
                } else {
                    // 设置箭头大小
                    resizeArrows(footer, pxValue)
                }
            }
        }
    }
    
    /**
     * 隐藏箭头
     */
    private fun hideArrows(footer: ClassicsFooter) {
        try {
            val arrowFieldNames = listOf("mArrowView", "mProgressView")
            var searchClass: Class<*>? = ClassicsFooter::class.java
            while (searchClass != null && searchClass != Any::class.java) {
                for (fieldName in arrowFieldNames) {
                    try {
                        val field = searchClass.getDeclaredField(fieldName)
                        field.isAccessible = true
                        val view = field.get(footer)
                        if (view is View) {
                            view.visibility = View.GONE
                        }
                    } catch (_: NoSuchFieldException) {}
                }
                searchClass = searchClass.superclass
            }
        } catch (_: Exception) {
            // 静默处理异常
        }
    }
    
    /**
     * 调整箭头大小
     */
    private fun resizeArrows(footer: ClassicsFooter, pxValue: Int) {
        try {
            val arrowFieldNames = listOf("mArrowView", "mProgressView")
            var searchClass: Class<*>? = ClassicsFooter::class.java
            while (searchClass != null && searchClass != Any::class.java) {
                for (fieldName in arrowFieldNames) {
                    try {
                        val field = searchClass.getDeclaredField(fieldName)
                        field.isAccessible = true
                        val view = field.get(footer)
                        if (view is View) {
                            val layoutParams = view.layoutParams
                            layoutParams.width = pxValue
                            layoutParams.height = pxValue
                            view.layoutParams = layoutParams
                        }
                    } catch (_: NoSuchFieldException) {}
                }
                searchClass = searchClass.superclass
            }
        } catch (_: Exception) {
            // 静默处理异常
        }
    }
    
    /**
     * 设置间距
     */
    private fun setMargin(footer: ClassicsFooter, props: Map<String, Any?>) {
        props["footerDrawableMarginRight"]?.let { 
            if (it is Number) {
                footer.setDrawableMarginRight(dpToPx(it.toDouble()))
            }
        }
    }
    
    /**
     * dp 转 px
     * 
     * 将密度无关像素（dp）转换为实际像素（px）
     * 
     * @param dp 密度无关像素值
     * @return 对应的像素值
     */
    private fun dpToPx(dp: Double): Float {
        val density = resources.displayMetrics.density
        return (dp * density).toFloat()
    }
}

File: ios/FooterPropsHandler.swift
import UIKit
import MJRefresh

/**
 * Footer 属性处理器
 * 
 * 负责处理 MJRefresh Footer 的所有属性设置
 * 包括：
 * - 文字内容（上拉提示、释放提示、加载中、没有更多数据等各种状态的文字）
 * - 颜色设置（强调色）
 * - 字体大小
 * - 箭头/进度图标的显示控制
 * - 其他配置
 * 
 * 注意：
 * - MJRefreshAutoFooter 没有 stateLabel，不支持自定义文字和样式
 * - MJRefreshBackNormalFooter 支持完整的文字和样式自定义
 */
class FooterPropsHandler {
    
    /**
     * 应用 Footer 属性到 MJRefreshBackNormalFooter
     * 
     * @param footer MJRefreshBackNormalFooter 实例
     * @param props 属性映射表
     */
    static func applyProps(footer: MJRefreshBackNormalFooter, props: [String: Any]) {
        // 设置文字内容
        setTextContent(footer: footer, props: props)
        
        // 设置颜色
        setColors(footer: footer, props: props)
        
        // 设置字体大小
        setTextSize(footer: footer, props: props)
        
        // 设置箭头显示
        setArrowVisibility(footer: footer, props: props)
        
        // 设置其他属性
        setOtherProps(footer: footer, props: props)
    }
    
    /**
     * 应用 Footer 属性到通用 MJRefreshFooter
     * 
     * 用于 MJRefreshAutoFooter 等不支持详细配置的 Footer
     * 
     * @param footer MJRefreshFooter 实例
     * @param props 属性映射表
     */
    static func applyGenericProps(footer: MJRefreshFooter, props: [String: Any]) {
        // MJRefreshAutoFooter 的样式配置较为有限（静默处理）
        _ = props["REFRESH_FOOTER_PULLING"]
        _ = props["REFRESH_FOOTER_LOADING"]
        _ = props["REFRESH_FOOTER_NOTHING"]
    }
    
    // MARK: - 私有方法
    
    /**
     * 设置文字内容
     * 
     * 设置 Footer 在不同状态下显示的文字：
     * - REFRESH_FOOTER_PULLING: 上拉过程中的提示文字
     * - REFRESH_FOOTER_RELEASE: 释放即可加载的提示文字
     * - REFRESH_FOOTER_LOADING: 加载中的提示文字
     * - REFRESH_FOOTER_NOTHING: 没有更多数据的提示文字
     * 
     * @param footer MJRefreshBackNormalFooter 实例
     * @param props 属性映射表
     */
    private static func setTextContent(footer: MJRefreshBackNormalFooter, props: [String: Any]) {
        if let pullingText = props["REFRESH_FOOTER_PULLING"] as? String {
            footer.setTitle(pullingText, for: .idle)
            footer.setTitle(pullingText, for: .pulling)
        }
        
        if let releaseText = props["REFRESH_FOOTER_RELEASE"] as? String {
            footer.setTitle(releaseText, for: .willRefresh)
        }
        
        if let loadingText = props["REFRESH_FOOTER_LOADING"] as? String {
            footer.setTitle(loadingText, for: .refreshing)
        }
        
        if let nothingText = props["REFRESH_FOOTER_NOTHING"] as? String {
            footer.setTitle(nothingText, for: .noMoreData)
        }
        
        // iOS MJRefresh 不支持的状态文字（静默处理）
        _ = props["REFRESH_FOOTER_REFRESHING"]
        _ = props["REFRESH_FOOTER_FINISH"]
        _ = props["REFRESH_FOOTER_FAILED"]
    }
    
    /**
     * 设置颜色
     * 
     * - footerAccentColor: 强调色（文字、箭头、进度条的颜色）
     * - footerPrimaryColor: 主色调（背景色，iOS MJRefresh 不支持）
     * 
     * @param footer MJRefreshBackNormalFooter 实例
     * @param props 属性映射表
     */
    private static func setColors(footer: MJRefreshBackNormalFooter, props: [String: Any]) {
        if let accentColor = props["footerAccentColor"] as? String,
           let color = UIColor(hexString: accentColor) {
            footer.stateLabel?.textColor = color
            // iOS 的 MJRefresh 箭头颜色与文字颜色绑定
            footer.arrowView?.tintColor = color
        }
        
        // iOS MJRefresh 不支持背景色设置
        _ = props["footerPrimaryColor"]
    }
    
    /**
     * 设置字体大小
     * 
     * - footerTitleTextSize: 标题文字大小（单位：pt）
     * 
     * @param footer MJRefreshBackNormalFooter 实例
     * @param props 属性映射表
     */
    private static func setTextSize(footer: MJRefreshBackNormalFooter, props: [String: Any]) {
        if let titleSize = props["footerTitleTextSize"] as? CGFloat {
            footer.stateLabel?.font = UIFont.systemFont(ofSize: titleSize)
        }
    }
    
    /**
     * 设置箭头显示
     * 
     * 支持两个属性（优先使用第一个）：
     * - footerDrawableArrowSize: 专门设置箭头大小
     * - footerDrawableSize: 通用图标大小设置
     * 
     * 特殊值：
     * - 0: 隐藏箭头
     * - 其他值: iOS MJRefresh 不支持动态调整箭头大小，仅记录日志
     * 
     * @param footer MJRefreshBackNormalFooter 实例
     * @param props 属性映射表
     */
    private static func setArrowVisibility(footer: MJRefreshBackNormalFooter, props: [String: Any]) {
        // 箭头大小设置（优先使用 footerDrawableArrowSize，回退到 footerDrawableSize）
        let arrowSize = props["footerDrawableArrowSize"] as? CGFloat ?? props["footerDrawableSize"] as? CGFloat
        
        if let size = arrowSize {
            // 如果设置为 0，隐藏箭头
            if size == 0 {
                footer.arrowView?.isHidden = true
            } else {
                footer.arrowView?.isHidden = false
                // iOS 的 MJRefresh 箭头大小是固定的，无法直接修改
            }
        }
        
        // iOS MJRefresh 不支持的属性（静默处理）
        _ = props["footerDrawableProgressSize"]
        _ = props["footerDrawableMarginRight"]
    }
    
    /**
     * 设置其他属性
     * 
     * - footerFinishDuration: 加载完成后停留时长（iOS MJRefresh 不支持）
     * 
     * @param footer MJRefreshBackNormalFooter 实例
     * @param props 属性映射表
     */
    private static func setOtherProps(footer: MJRefreshBackNormalFooter, props: [String: Any]) {
        // iOS MJRefresh 不支持的属性（静默处理）
        _ = props["footerFinishDuration"]
    }
}
